Block serialization compression

kchadha opened the pull request:
Resolves #22 
Resolves LLK/scratch-vm#194

#### Proposed Changes
This pull request is one of the last major changes for the save/load work and should solidify the 3.0 format. Specifically, the changes made here cover the pre-compression work for the Scratch 3.0 format (related to LLK/scratch-vm#).

#### Related Pull Requests
This pull request is tied to the following others in scratch-vm and scratch-gui and should only be merged when all are ready since there are dependencies in both directions.

#### Testing
Testing has been done manually on importing 2.0 projects into 3.0 with 2 rounds of saving and loading. Actual fixtures to be added in the near future for the 3.0 format. Tracking this as an issue in #31.

kchadha commented:
@cwillisf I think it would be good to get your review on this PR too before it gets pulled in

rschamp commented on lib/sb3_schema.json:
```diff
@@ -231,8 +296,22 @@
+                },
+                "videoState": {
+                    "type": "string",
+                    "enum": ["on", "off", "on_flipped"]
```
This should use a hyphen since it's what we display to the user.

cwillisf commented on lib/sb3_schema.json:
```diff
@@ -29,7 +41,7 @@
                 "dataFormat": {
                     "type": "string",
-                    "enum": ["png", "PNG", "svg", "SVG", "jpeg", "JPEG", "jpg", "JPG", "bmp", "BMP"]
+                    "enum": ["png", "PNG", "svg", "SVG", "jpeg", "JPEG", "jpg", "JPG", "bmp", "BMP", "gif", "GIF"]
```
Is it possible to set these to ignore capitalization? Or maybe something like:
```
"pattern": ["[pP][nN][gG]", "[sS][vV][gG]", ...
```

It's not a big deal, but I have on occasion seen files with extensions like "Jpeg." Maybe we should assume that the editor (GUI or whatever) will fix up the file name before saving, though... in which case maybe we should just list out the lower-case versions?

kchadha replied:
Good point! I think I'll change it to only accept lower case and make sure that we're only serializing lowercase.

cwillisf commented on lib/sb3_schema.json:
```diff
@@ -87,98 +99,152 @@
+                        "items": [
+                            {
+                                "type":"number",
+                                "enum":[1,2,3]
```
~Are these type numbers documented somewhere?~
I see that they're documented in LLK/scratch-vm#1031 but it would be nice to have them documented in this repository as well, somehow.

kchadha commented:
@cwillisf, @rschamp, @thisandagain, made some changes addressing PR comments that were made and also undoing the validation workaround I had for the scratch-blocks/scratch-vm bug where orhpan shadow blocks get created for custom procedure calls every time there is a pass through xml (LLK/scratch-vm#1011). This will be handled in VM instead (look out for a change coming soon to that PR.

cwillisf commented on lib/sb3_schema.json:
```diff
@@ -71,15 +71,15 @@
                 "md5ext": {
                     "type": "string",
-                    "pattern": "^[a-fA-F0-9]{32}.[a-zA-Z]+$"
+                    "pattern": "^[a-fA-F0-9]{32}.[a-zA-Z0-9]+$"
```
Ha! Good catch ;)

tjvr replied:
Sorry to interrupt but shouldn't the dot be escaped, if this is a regex? :-) `\.`

kchadha replied:
good catch, @tjvr!

cwillisf commented on lib/sb3_schema.json:
```diff
@@ -129,6 +129,7 @@
                 {
                     "type": "number",
                     "enum": [4,5,6,7,8]
+
```
👎

cwillisf commented on lib/sb3_schema.json:
```diff
@@ -232,7 +233,8 @@
                                 "type":"number",
-                                "enum":[1,2,3]
+                                "enum":[1,2,3],
+                                "description": "1 = unobscured shadow, 2 = no shadow, 3 = obscured shadow"
```
👍
